### ISSUE TITLE
Bug: [@typescript-eslint/no-unused-vars] <wrong report of unused param in generics>

### ISSUE THREAD

**Issue by fdc-viktor-luft** [typescript-eslint/typescript-eslint#7211]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.1.6&fileType=.ts&code=FAFwngDgpgBA6gSxACwOJQHZQE4IMYA8AKgHwwC8MRA3KJLAGID2TFMA3jABRQBcMAZxC4MAcwCU-AG5MEAExgBfajAD0qmBlYB3AIbYMCMXWgxmTAIxtEKdFlyFzJWuFPmATNaRpMOfAU4efiERCWlZBUVnNQ09AyNRYGAgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUFvcgAQBcBPABxQGNoBLEggWhXioDsCB6ZgeztmdmUQATOgDcAhtFQYA2uBzzIAdwnNIc%2BWAC6cgL4gdQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkeAhpgObp4VTokBMADL5ABpwNSPFgATAK4Z2kTkXQAPNkJpQSeTMmhtKmyelUiAViUWzEmTlgBu6QcNrIksTHgAS6VIixk9mAyNaMSkMACVOWFRJAmJZIgl7IPUAa2REABlkACMAYQALdGgU9n1DRw4SAFkJaXQASSI8LFhEUoDytUh6VFRYAHcAZQBPJsKCaAARdAAzemk8epc3PzAy5MgSWElMaHQq%2Bjb-QIrIKy3UOwApLaIa0PtjzpEEgFF4ZF01jo3UHPYANoVWjyJQqYFQcQIBxqWhQ%2BAAOk%2BWHo2RkyVoJGgWnw%2BVgGkgFQAusIAL4gUlAA&tokens=false

### Repro Code

```typescript
type WithGeneric<T> = T;
type Foo = { (e: string): void }; // no warning
type Foo1 = WithGeneric<Foo>;
type Foo2 = WithGeneric<{ (e: string): void }>; // warning
```

### ESLint Config

```javascript
module.exports = {
  extends: [
        "eslint:recommended",
        "plugin:vue/vue3-essential",
        "plugin:@intlify/vue-i18n/recommended",
        "@vue/eslint-config-typescript/recommended",
        "@vue/eslint-config-prettier",
    ],
    plugins: ["import", "simple-import-sort"],
    parserOptions: {
        ecmaVersion: "latest",
    },
    rules: {
        "@typescript-eslint/no-unused-vars": ["warn"],
    },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "target": "es2020",
    "module": "esnext",
    "strict": true,
    "jsx": "preserve",
    "importHelpers": true,
    "moduleResolution": "node",
    "skipLibCheck": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "sourceMap": true,
    "resolveJsonModule": true,
    "noEmit": true,
    "lib": [
      "esnext",
      "dom",
      "dom.iterable",
      "scripthost"
    ]
  }
}
```

### Expected Result

No warnings for above code. It should not matter if the type is declared using `type` or inline in a generic placeholder.

### Actual Result

I get only no warning if I declare an extra type.

### Additional Info

<img width="435" alt="Bildschirmfoto 2023-07-12 um 17 49 43" src="https://github.com/typescript-eslint/typescript-eslint/assets/17571883/6ed1c2db-bde0-46f0-ab90-dd92353a453f">

I could not reproduce the issue on your playground, so I expect here some side effects from other rules if that is possible.

**Comment by bradzacher** [comment]
> I could not reproduce the issue on your playground, so I expect here some side effects from other rules if that is possible.

Without an isolated repro it's nigh impossible for us to look into this for you.
Could be incorrect config or out of date versions - hard to say.

I'm going to close this as no repro but feel free to comment back if you can isolate a repro. Feel free to create a separate GH repo and link it here.

**Comment by fdc-viktor-luft** [comment]
I can confirm now that it happens only after the v6.0.0 update and works without issues using v5.62.0.

All other dependencies are up-to-date. But I'll try to find some time creating a repro repo and then come back to you.

**Comment by fdc-viktor-luft** [comment]
Update: Still happens in v6.1.0

**Comment by bradzacher** [comment]
As soon as you have an isolated repro we can investigate! Until then it's unlikely something will change.

**Comment by fdc-viktor-luft** [comment]
Here it is: [#typescript-eslint-issue-7211](https://github.com/fdc-viktor-luft/typescript-eslint-issue-7211)

**Comment by bradzacher** [comment]
Fix for your config:
```js
module.exports = {
  // ...
  overrides: [
    // most basic
    {
      files: ['*.ts', '*.tsx'],
      parser: '@typescript-eslint/parser',
    },
    // or even better
    {
      files: ['*.ts', '*.tsx'],
      extends: ['plugin:@typescript-eslint/recommended'],
    }
  ]
};
```

In a nutshell you're not directly using our parser - you're relying on `vue-eslint-parser` to parse TS files and it calls out to our parser - only it's not correctly calling our tooling.
The above config overrides things so that you're using our parser for ts files.

I'd consider filing a bug with them.

**Comment by fdc-viktor-luft** [comment]
Will do thanks :+1:
